File: src/main/java/io/primeaspect/calculator/service/enums/Delimiter.java
package io.primeaspect.calculator.service.enums;

import java.util.Arrays;

public enum Delimiter {
    LEFT("(") {
        public int getPriority() {
            return 1;
        }
    },
    RIGHT(")") {
        public int getPriority() {
            return 5;
        }
    };

    private final String value;
    private static String operations;

    Delimiter(String value) {
        this.value = value;
    }

    public String getValue() {
        return value;
    }

    public abstract int getPriority();

    public static boolean isDelimiter(String token) {
        return getDelimiters().contains(token);
    }

    public static Delimiter getDelimiter(String token) {
        return token.equals("(") ? LEFT : RIGHT;
    }

    public static String getDelimiters() {
        if (operations == null) {
            StringBuilder result = new StringBuilder();
            Arrays.asList(Delimiter.values()).forEach(value -> result.append(value.getValue()));
            operations = result.toString();
        }
        return operations;
    }
}

File: src/main/kotlin/io/primeaspect/calculator/service/enums/Delimiter.kt
package io.primeaspect.calculator.service.enums

enum class Delimiter(val value: String, val priority: Int) {
    LEFT("(", 1),
    RIGHT(")", 5);

    companion object {
        private var operations: String? = null

        fun isDelimiter(token: String?) = getDelimiters().contains(token!!)

        fun getDelimiter(token: String) = when (token) {
            "(" -> LEFT
            else -> RIGHT
        }

        fun getDelimiters(): String {
            if (operations == null) {
                val result = StringBuilder()
                values().forEach {
                    result.append(it.value)
                }
                operations = result.toString()
            }
            return operations!!
        }
    }
}